test/io/request_test.dart

```
@@ -10,61 +10,48 @@ import 'package:test/test.dart';
+    final bytesString = await response.stream.bytesToString();
     expect(
-        startServer().then((_) {
-          var request = http.Request('POST', serverUrl.resolve('/redirect'))
-            ..followRedirects = false;
-          var future = request.send().then((response) {
-            expect(response.statusCode, equals(302));
-          });
-          expect(
-              future.catchError((_) {}).then((_) => stopServer()), completes);
-          expect(future, completes);
-        }),
-        completes);
+        bytesString,
+        parse(equals({
+          'method': 'GET',
+          'path': '/',
+          'headers': {
+            'content-type': ['text/plain; charset=utf-8'],
+            'accept-encoding': ['gzip'],
+            'user-agent': ['Dart'],
+            'content-length': ['5']
+          },
+          'body': 'hello',
+        })));
+  });
+
+  test('#followRedirects', () async {
+    final request = http.Request('GET', serverUrl.resolve('/redirect'))
+      ..followRedirects = false;
+    final response = await request.send();
+
+    expect(response.statusCode, equals(302));
   });
 
-  test('#maxRedirects', () {
+  test('#maxRedirects', () async {
+    final request = http.Request('GET', serverUrl.resolve('/loop?1'))
+      ..maxRedirects = 2;
     expect(
-        startServer().then((_) {
-          var request = http.Request('POST', serverUrl.resolve('/loop?1'))
-            ..maxRedirects = 2;
-          var future = request.send().catchError((error) {
-            expect(error, isRedirectLimitExceededException);
-            expect(error.redirects.length, equals(2));
-          });
-          expect(
-              future.catchError((_) {}).then((_) => stopServer()), completes);
-          expect(future, completes);
-        }),
-        completes);
+        request.send(),
+        throwsA(isA<http.ClientException>()
+            .having((e) => e.message, 'message', 'Redirect limit exceeded')));
```

This used to also test that the number of redirects was correct, which is worth restoring

---

`package:http` doesn't make that possible, we hide the original exception and keep only it's message.

We could add a `Object originalException` to `ClientException`, but it would be a new feature and we'd want to think it through.

---

Note that I did call this out in the commit message 😉 

> We also lose the underlying
> exception and have only the message so we can't check the length of
> the `redirects` field.

---

So was this a problem before as well? Like the test before never would have gotten the expected error type?

---

I suppose you could check on the server side how many requests came through but likely not worth doing.

---

> So was this a problem before as well? Like the test before never would have gotten the expected error type?

Yes, previously adding a `expectAsync1` around the error callback would make the test time out. Fixing the request type to `'GET'` would then make it fail the first `expect` because it gets a `HttpClientException` and not a `RedirectException`.

---

:man_facepalming: